core/sync/index.js

```diff
@@ -231,28 +364,217 @@ class Sync {
+          case remoteErrors.FILE_TOO_LARGE_CODE:
+          case remoteErrors.INVALID_FOLDER_MOVE_CODE:
+          case remoteErrors.INVALID_METADATA_CODE:
+          case remoteErrors.INVALID_NAME_CODE:
+          case remoteErrors.NEEDS_REMOTE_MERGE_CODE:
+          case remoteErrors.NO_COZY_SPACE_CODE:
+          case remoteErrors.PATH_TOO_DEEP_CODE:
+          case remoteErrors.UNKNOWN_INVALID_DATA_ERROR_CODE:
+          case remoteErrors.UNKNOWN_REMOTE_ERROR_CODE:
+          case remoteErrors.UNREACHABLE_COZY_CODE:
+          case remoteErrors.USER_ACTION_REQUIRED_CODE:
+            // We will keep retrying to apply the change until it's fixed or the
+            // user contacts our support.
+            // See `default` case for other blocking errors for which we'll stop
+            // retrying after 3 failed attempts.
+            this.blockSyncFor({ err, change })
+            break
+          case remoteErrors.CONFLICTING_NAME_CODE:
+            /* When we fail to apply a change because of a name conflict on the
+             * remote Cozy, it means we either:
+             * 1. have another change to apply that will free the give name by
+             *    moving the document or renaming it
+             * 2. have not yet merged the remote change that took the name and
+             *    would have generated a conflict copy
+             * 3. have not merged the remote change that took the name and never
+             *    will because we abandoned that change in the past
+             *
+             * We can solve 1. by marking our local change with an increased
+             * error counter so it can be retried later, after we've applied the
+             * other changes that would free the conflicting name.
+             *
+             * We can solve 2. by blocking the Sync process until we've fetched
+             * and merged the new remote changes, thus generating a conflict
+             * copy at the Merge level.
+             *
+             * We can only solve 3. by detecting that we've tried the solutions
+             * to 1. and 2., still can't apply the given change and thus generate
+             * a remote conflict at the Sync level instead of doing it at the
+             * Merge level.
+             */
+
+            if (shouldAttemptRetry(change)) {
+              // Solve 1. & 2.
+              this.blockSyncFor({ err, change })
+            } else {
+              log.error(
+                { path, err, change },
+                'Document already exists on Cozy'
+              )
+              // Solve 3.
+              // We resolve the conflict on the remote Cozy to avoid local
+              // issues like permission issues on Windows.
+              await this.remote.resolveRemoteConflict(change.doc)
+            }
+            break
+          case remoteErrors.DOCUMENT_IN_TRASH_CODE:
+            delete change.doc.moveFrom
+            delete change.doc.overwrite
```

Comment:
[nit] I would have kept the `overwrite` in case trashing the file was a mistake and the user restore the file just after. We probably don't really care as it would be very rare to have the good conditions and timing to happen, but if keeping the `overwrite` doesn't cost us anything, I still think it would be more "correct".

Reply:
I'm not sure to follow. If we remove the `moveFrom` attribute then the `overwrite` attribute will never be used by the Writers so I don't see any good reason to keep it.
On the other hand, I'm pretty sure keeping it could lead to unexpected behaviors.